src/filesystem_littlefs.c

```diff
@@ -0,0 +1,517 @@
+        fsync_r(r, fileStruct);
+    }
+
+    return result;
+}
+
+static ssize_t read_r(struct _reent *r, void *fileStruct, char *ptr, size_t len) {
+    PREPARE_FIL_EX(f, fileStruct);
+    int result;
+
+    if (!ptr) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    if (len == 0) {
+        r->_errno = 0;
+        return 0;
+    }
+
+    result = lfs_file_read(lfs, &(f->fil), ptr, len);
+    r->_errno = get_errno(result);
+
+    if (result < 0) {
+        return -1;
+    }
+
+    return result;
+}
+
+static off_t seek_r(struct _reent *r, void *fileStruct, off_t pos, int dir) {
+    PREPARE_FIL_EX(f, fileStruct);
+    int result;
+
+    result = lfs_file_seek(lfs, &(f->fil), pos, dir);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int stat_r(struct _reent *r, const char *path, struct stat *st) {
+    int result;
+    struct lfs_info info;
+    memset(&info, 0, sizeof(struct lfs_info));
+    memset(st, 0, sizeof(struct stat));
+
+    if (IS_EMPTY(path)) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    path = process_path(path, devoptab_littlefs.name);
+
+    result = lfs_stat(lfs, path, &info);
+        r->_errno = get_errno(result);
+
+    if (result < 0) {
+        return -1;
+    }
+
+    st->st_mode = ALLPERMS; // Permissions handling not implemented, use all
+
+    if (info.type & LFS_TYPE_DIR) {
+        st->st_mode = S_IFDIR;
+    } else if (info.type & LFS_TYPE_REG) {
+        st->st_mode = S_IFREG;
+    } else {
+        // Unexpected type, return error
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    st->st_dev = device;
+    st->st_size = info.size;
+    st->st_blksize = lfs->cfg->block_size;
+    st->st_blocks = 1;
+
+    return 0;
+}
+
+static int fstat_r(struct _reent *r, void *fileStruct, struct stat *st) {
+    PREPARE_FIL_EX(f, fileStruct);
+    memset(st, 0, sizeof(struct stat));
+
+    st->st_mode = ALLPERMS; // Permissions handling not implemented, use all
+
+    if (f->fil.type & LFS_TYPE_DIR) {
+        st->st_mode = S_IFDIR;
+    } else if (f->fil.type & LFS_TYPE_REG) {
+        st->st_mode = S_IFREG;
+    } else {
+        // Unexpected type, return error
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    st->st_dev = device;
+    st->st_size = f->fil.ctz.size;
+    st->st_blksize = lfs->cfg->block_size;
+    st->st_blocks = 1;
+
+    return 0;
+}
+
+static int link_r(struct _reent *r, const char *existing, const char *newLink) {
+    // Links are not supported on littlefs
+    r->_errno = ENOTSUP;
+    return -1;
+}
+
+static int unlink_r(struct _reent *r, const char *path) {
+    int result;
+
+    if (IS_EMPTY(path)) {
+        r->_errno = EINVAL;
+        return -1;
+    }
+
+    path = process_path(path, devoptab_littlefs.name);
+
+    result = lfs_remove(lfs, path);
+    r->_errno = get_errno(result);
+
+    return RESULT;
+}
+
+static int chdir_r(struct _reent *r, const char *path) {
+    // chdir not implemented for littlefs
```

Comment:
no support for changing current directory, that's sad :smile: (just joking)

Reply:
As littlefs doesn't support `chdir` and I don't think we will needed it, I didn't spend time implementing it :) If needed, it would require just storing the path set in this function and modifying `process_path` function.